### PR TITLE
I103 redesign edit to be a modal in visit

### LINKED ISSUE #103: Redesign edit to be a modal in visit
## Basic Information

Redesign edit to be a modal in visit, looks pre cool but yeah people want it changed. 
Also change display name to client name, and have pets be pulled from contacts.

- [x] remove current
- [x] add modal
- [x] change display name
- [x] pull pets from contacts
- [x] works

## Other Information
:))

Potential scopes:

### PR DESCRIPTION
**Pull request by Dustybluelight**
## Change Summary
- The visit edit now sends you to the same visit form which is now on the `/visit/set` page.
- Client's names and pets are drawn from the contact list and saved in the visit.
- `displayName` is now `clientName`
- End time is now a duration selector in increments of 15 minutes
- Add alerts for form submission
- Add back mysterious missing margin on visit instances
- Moved `/components/Visit/utils.ts` to `/utils/index.ts`

:+1: 

# Related Issue

- Resolves #103 and #111

### CODE REVIEW

**Comment by dct0** [comment]
Should reusable button be integrated in this issue as well?

**Comment by scratchclaggy** [comment]
> Should reusable button be integrated in this issue as well?
@dct0 
If you're happy to use it, you can merge `i63-reusable-button` into this branch and that should be ready to go (pending review). Otherwise, if this PR goes through before #132 then I'll swap out the buttons.

File: src/components/Visit/formfield.tsx [hunk 1]
```diff
@@ -1,12 +1,14 @@
-import { ChangeEvent } from 'react'
+import React, { ChangeEvent } from 'react'
```
Comment (skyheat) [inline src/components/Visit/formfield.tsx:1]: Importing react is unnecessary as next.js automatically imports it.

**Review by skyheat — CHANGES_REQUESTED** [review]
No glaring issues from what I can see.

Good work!

**Review by scratchclaggy — COMMENTED** [review]
Good stuff so far, I've left a few comments throughout each of the files.

Some additional things:

image
Is it possible to not use the US date format with this date picker component?

File: src/components/Visit/utils.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    'Friday',
+    'Saturday'
+  ]
+  const date = timestamp.toDate()
+  return `${days[date.getDay()]}, ${date.toLocaleString().slice(0, -3)}`
+}
+
+export const findContactIndex = (id: string, userDoc: UserData) => {
+  for (let i = 0; i < userDoc.contacts.length; i++) {
+    if (userDoc.contacts[i].id === id) {
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/components/Visit/utils.ts:42]: Could grab the contact's info by the name instead?

File: public/sw.js.map [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (dct0) [reply public/sw.js.map:1]: Looks like an old file that should be ignored in the .gitignore. Deleted.

File: src/pages/visit/set.tsx [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            <FormField
+              className='col-span-2'
+              id='startTimeInput'
+              type='dateTime-local'
+              placeholder='Start Time'
+              value={startTime}
+              label='Start Time:'
+              isRequired={true}
+              onChange={(event) => {
+                // event.target.value = event.target.value.substring(0, 16) // fixes invalid input on ios safari? can't test
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/pages/visit/set.tsx:161]: Uncommented. Maybe someone with mobile safari can check this works?

File: src/components/Visit/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (NicholasChoong) [reply src/components/Visit/utils.ts:42]: > @NicholasChoong I guess we need to keep the `id` field?

yea

[the remark this answers is not in the document]
Comment (scratchclaggy) [reply src/components/Visit/utils.ts:42]: In the contacts page, we're finding contacts via index, so Nic and I were confused about why contacts needed an `id` as well. Is it possible to use the same technique here?

File: src/components/Visit/utils.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  }
+  return parseFloat(value)
+}
+
+export const padNumber = (value: number) => {
+  return value.toString().padStart(2, '0')
+}
+
+// Returns YYYY-MM-DDTHH:mm for input value
+export const formatTimestamp = (timestamp?: Timestamp) => {
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/components/Visit/utils.ts:17]: It seems like ` Intl.DateTimeFormat()` requires a specified locale. If none is specified, it does not add the time so I think `humanizeTimestamp()` is fine.
```js
const date = new Date(Date.UTC(2020, 11, 20, 3, 23, 16, 738));

console.log(new Intl.DateTimeFormat('en-AU', { dateStyle: 'long', timeStyle: 'medium', timeZone: 'Australia/Sydney' }).format(date));
// Output: "20 December 2020 at 2:23:16 pm"

console.log(new Intl.DateTimeFormat({ dateStyle: 'long', timeStyle: 'medium', timeZone: 'Australia/Sydney' }).format(date))
// Output: "20/12/2020" on my browser
```

File: src/pages/visit/set.tsx [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  const handleSubmit = (click: React.FormEvent<HTMLFormElement>) => {
+    click.preventDefault()
+
+    const data: VisitData = {
+      type: visitType,
+      clientId: clientId, // possible refac: push a client object
+      clientName: clientName,
+      startTime: Timestamp.fromDate(new Date(startTime)),
+      duration: duration,
+      petNames: userDoc.contacts[findContactIndex(clientName, userDoc)].pets, // TODO?: find contact by id instead
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/pages/visit/set.tsx:56]: Removed todos

File: src/pages/visit/set.tsx [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      walkDist: walkDist,
+      commuteDist: commuteDist,
+      commuteMethod: commuteMethod,
+      notes: notes
+    }
+    if (id !== null) {
+      userDoc.visits[id] = data
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/pages/visit/set.tsx:63]: I think I did it?

File: src/components/Visit/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dct0) [reply src/components/Visit/utils.ts:42]: Removed the function.

File: src/components/Visit/utils.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
+import { Timestamp } from 'firebase/firestore'
```
[the remark this answers is not in the document]
Comment (dct0) [reply src/components/Visit/utils.ts]: Moved file to `src/utils/index.ts`

**Review by skyheat — APPROVED** [review]
Nice

**Pull request merged**